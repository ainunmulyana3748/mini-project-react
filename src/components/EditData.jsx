import InputForm from "../components/InputForm";
import SuccessModal from "../components/SuccessModal";
import Button from "../components/Button";
import { X } from "lucide-react";
import { useState } from "react";
import ConfirmModal from "./ConfirmModal";
import { useNavigate } from "react-router-dom";

const EditData = ({ onclick, user }) => {
  const [showConfirmEditModal, setShowConfirmEditModal] = useState(false);
  const [showSuccessModal, setShowSuccessModal] = useState(false);
  const navigate = useNavigate("/");

  const [formData, setFormData] = useState({
    first_name: user.first_name || "",
    last_name: user.last_name || "",
    email: user.email || "",
  });

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  return (
    <div className="mt-10 w-full flex items-center justify-center">
      <div className="w-full max-w-xl mx-auto bg-white rounded-3xl shadow-lg p-6 sm:p-8 mt-6 transition-all duration-300 hover:shadow-blue-100">
        <div className="flex justify-end">
          <X
            size={28}
            className="cursor-pointer text-gray-600 hover:text-blue-600 transition"
            onClick={onclick}
          />
        </div>

        <form className="space-y-4">
          <InputForm
            label="First Name"
            type="text"
            name="first_name"
            value={formData.first_name}
            onChange={handleChange}
            placeholder="John"
          />
          <InputForm
            label="Last Name"
            type="text"
            name="last_name"
            value={formData.last_name}
            onChange={handleChange}
            placeholder="Doe"
          />
          <InputForm
            label="Email"
            type="email"
            name="email"
            value={formData.email}
            onChange={handleChange}
            placeholder="john@example.com"
          />
          <Button
            variant="primary"
            onclick={(e) => {
              e.preventDefault();
              setShowConfirmEditModal(true);
            }}
          >
            Edit
          </Button>
        </form>
      </div>

      {showConfirmEditModal && (
        <ConfirmModal
          type="edit"
          title="Confirm Changes"
          message="Are you sure you want to save the changes to this guide?"
          onConfirm={() => {
            setShowSuccessModal(true);
            setTimeout(() => {
              setShowSuccessModal(false);
              navigate("/");
            }, 2000);
          }}
          onCancel={() => setShowConfirmEditModal(false)}
        />
      )}

      {showSuccessModal && (
        <SuccessModal
          type="edit"
          title="Edit Successful!"
          message="The guide profile has been updated."
        />
      )}
    </div>
  );
};

export default EditData;
