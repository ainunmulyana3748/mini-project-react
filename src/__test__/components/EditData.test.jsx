import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import EditData from "../../components/EditData";
import { BrowserRouter } from "react-router-dom";

// Mock useNavigate
const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("EditData Component", () => {
  const user = {
    first_name: "John",
    last_name: "Doe",
    email: "john@example.com",
  };
  const onClose = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("render form dengan data user", () => {
    render(
      <BrowserRouter>
        <EditData user={user} onclick={onClose} />
      </BrowserRouter>
    );
    expect(screen.getByDisplayValue("John")).toBeInTheDocument();
    expect(screen.getByDisplayValue("Doe")).toBeInTheDocument();
    expect(screen.getByDisplayValue("john@example.com")).toBeInTheDocument();
  });

  test("ubah input dan klik Edit muncul ConfirmModal", () => {
    render(
      <BrowserRouter>
        <EditData user={user} onclick={onClose} />
      </BrowserRouter>
    );
    fireEvent.change(screen.getByPlaceholderText("John"), {
      target: { value: "Jane" },
    });
    fireEvent.click(screen.getByText("Edit"));
    expect(screen.getByText("Confirm Changes")).toBeInTheDocument();
  });

  test("klik Cancel di ConfirmModal menutup modal", () => {
    render(
      <BrowserRouter>
        <EditData user={user} onclick={onClose} />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByText("Edit"));
    fireEvent.click(screen.getByText("Cancel"));
    expect(screen.queryByText("Confirm Changes")).not.toBeInTheDocument();
  });

  test("klik Confirm di ConfirmModal muncul SuccessModal dan redirect", async () => {
    jest.useFakeTimers();
    render(
      <BrowserRouter>
        <EditData user={user} onclick={onClose} />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByText("Edit"));
    fireEvent.click(screen.getByText("Yes, Save"));
    expect(screen.getByText("Edit Successful!")).toBeInTheDocument();
    // Simulasi timeout
    jest.runAllTimers();
    await waitFor(() => {
      expect(mockedUsedNavigate).toHaveBeenCalledWith("/");
    });
    jest.useRealTimers();
  });

  test("klik tombol close (X) memanggil onclick", () => {
    render(
      <BrowserRouter>
        <EditData user={user} onclick={onClose} />
      </BrowserRouter>
    );
    const closeBtn = screen.getByRole("img", { hidden: true });
    fireEvent.click(closeBtn);
    expect(onClose).toHaveBeenCalled();
  });

  test("snapshot edit data", () => {
    const { container } = render(
      <BrowserRouter>
        <EditData user={user} onclick={onClose} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
