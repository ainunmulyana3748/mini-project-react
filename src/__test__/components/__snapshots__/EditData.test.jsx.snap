// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditData Component snapshot edit data 1`] = `
<div>
  <div
    class="mt-10 w-full flex items-center justify-center"
  >
    <div
      class="w-full max-w-xl mx-auto bg-white rounded-3xl shadow-lg p-6 sm:p-8 mt-6 transition-all duration-300 hover:shadow-blue-100"
    >
      <div
        class="flex justify-end"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-x cursor-pointer text-gray-600 hover:text-blue-600 transition"
          fill="none"
          height="28"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </div>
      <form
        class="space-y-4"
      >
        <div
          class="space-y-1"
        >
          <label
            class="block text-sm font-medium"
            for="first-name"
          >
            First Name
          </label>
          <input
            class="w-full px-3 py-2 border border-zinc-300 rounded-md focus:outline-none focus:ring focus:ring-blue-300"
            id="first-name"
            name="first_name"
            placeholder="John"
            type="text"
            value="John"
          />
        </div>
        <div
          class="space-y-1"
        >
          <label
            class="block text-sm font-medium"
            for="last-name"
          >
            Last Name
          </label>
          <input
            class="w-full px-3 py-2 border border-zinc-300 rounded-md focus:outline-none focus:ring focus:ring-blue-300"
            id="last-name"
            name="last_name"
            placeholder="Doe"
            type="text"
            value="Doe"
          />
        </div>
        <div
          class="space-y-1"
        >
          <label
            class="block text-sm font-medium"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-3 py-2 border border-zinc-300 rounded-md focus:outline-none focus:ring focus:ring-blue-300"
            id="email"
            name="email"
            placeholder="john@example.com"
            type="email"
            value="john@example.com"
          />
        </div>
        <button
          class="px-4 py-2 rounded font-medium bg-black text-white hover:bg-zinc-900 focus:ring-black disabled:opacity-50 cursor-pointer"
        >
          Edit
        </button>
      </form>
    </div>
  </div>
</div>
`;
